src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Automata.cs

```diff
@@ -0,0 +1,429 @@
+        /// via https://github.com/dotnet/runtime/issues/63474), and use of Volatile.Read is
+        /// contributing non-trivial overhead (https://github.com/dotnet/runtime/issues/65789).
+        /// </remarks>
+        private int[] _delta;
+
+        /// <summary>
+        /// Maps each NFA state id to the state id of the DfaMatchingState stored in _stateArray.
+        /// This map is used to compactly represent NFA state ids in NFA mode in order to utilize
+        /// the property that all NFA states are small integers in one interval.
+        /// The valid entries are 0 to <see cref="NfaStateCount"/>-1.
+        /// </summary>
+        private int[] _nfaStateArray = Array.Empty<int>();
+
+        /// <summary>
+        /// Maps the id of a DfaMatchingState to the NFA state id that it is being identifed with in the NFA.
+        /// It is the inverse of used entries in _nfaStateArray.
+        /// The range of this map is 0 to <see cref="NfaStateCount"/>-1.
+        /// </summary>
+        private readonly Dictionary<int, int> _nfaStateArrayInverse = new();
+
+        /// <summary>Gets <see cref="_nfaStateArrayInverse"/>.Count</summary>
+        private int NfaStateCount => _nfaStateArrayInverse.Count;
+
+        /// <summary>
+        /// Transition function for NFA transitions in NFA mode.
+        /// Each NFA entry maps to a list of NFA target states.
+        /// Each list of target states is without repetitions.
+        /// If the entry is null then the targets states have not been computed yet.
+        /// </summary>
+        private int[]?[] _nfaDelta = Array.Empty<int[]>();
+
+        /// <summary>
+        /// The transition function for <see cref="FindSubcaptures(ReadOnlySpan{char}, int, int, PerThreadData)"/>,
+        /// which is an NFA mode with additional state to track capture start and end positions.
+        /// Each entry is an array of pairs of target state and effects to be applied when taking the transition.
+        /// If the entry is null then the transition has not been computed yet.
+        /// </summary>
+        private (int, DerivativeEffect[])[]?[] _capturingNfaDelta = Array.Empty<(int, DerivativeEffect[])[]?>();
+
+        /// <summary>
+        /// Implements a version of <see cref="Array.Resize"/> that is guaranteed to not publish an array before values
+        /// have been copied over.
+        /// </summary>
+        /// <remarks>
+        /// This may not be strictly necessary for arrays of primitive or reference types (which have atomic
+        /// reads/writes), as when, e.g., <see cref="_delta"/> is found to not have an entry the array is checked again
+        /// after a lock on the matcher has been acquired. However, in a highly threaded use case it still seems better
+        /// to avoid unnecessarily causing other threads to acquire the lock.
+        /// </remarks>
+        private static void ConcurrentArrayResize<T>(ref T[] array, int newSize)
```

Comment:
Right, so there are no concurrent calls to `ConcurrentArrayResize`, but there can be concurrent reads from the array that's being resized (other matchers taking transitions). My concern with using `Array.Resize` is that a valid implementation of it could first assign the pointer to the new array and only after that copy the data over, in which case those concurrent readers would see the copy in flight. The aim here is to make sure the whole copy operation is done before the new array is published to those concurrent reader threads.

Reply:
> valid implementation of it could first assign the pointer to the new array and only after that copy the data over,

Got it. That's a fair argument. We wouldn't do that, both because it would break code and because the docs state the order "This method allocates a new array with the specified size, copies elements from the old array to the new one, and then replaces the old array with the new one", but it's fine to be cautious. We should choose a different name, though, like the one I suggested later.

Reply:
I did read that sentence, but without the Volatile.Write I'd be concerned that the compiler/JIT/CPU might reorder the pointer assignment to before some of the copies. That wouldn't be a big deal right now since I think all the arrays have elements that write atomically, but indeed just to be cautious.